server/src/main/java/org/opensearch/common/cache/stats/MultiDimensionCacheStats.java

```diff
@@ -0,0 +1,265 @@
+    }
+
+    @Override
+    public long getTotalEntries() {
+        return getTotalStats().getEntries();
+    }
+
+    /**
+     * Returns a new tree containing the stats aggregated by the levels passed in. The root node is a dummy node,
+     * whose name and value are null. The new tree only has dimensions matching the levels passed in.
+     */
+    MDCSDimensionNode aggregateByLevels(List<String> levels) {
+        List<String> filteredLevels = filterLevels(levels);
+        MDCSDimensionNode newRoot = new MDCSDimensionNode("", true, statsRoot.getStats());
+        for (MDCSDimensionNode child : statsRoot.children.values()) {
+            aggregateByLevelsHelper(newRoot, child, filteredLevels, 0);
+        }
+        return newRoot;
+    }
+
+    void aggregateByLevelsHelper(
+        MDCSDimensionNode parentInNewTree,
+        MDCSDimensionNode currentInOriginalTree,
+        List<String> levels,
+        int depth
+    ) {
+        if (levels.contains(dimensionNames.get(depth))) {
+            // If this node is in a level we want to aggregate, create a new dimension node with the same value and stats, and connect it to
+            // the last parent node in the new tree. If it already exists, increment it instead.
+            String dimensionValue = currentInOriginalTree.getDimensionValue();
+            MDCSDimensionNode nodeInNewTree = parentInNewTree.children.get(dimensionValue);
+            if (nodeInNewTree == null) {
+                // Create new node with stats matching the node from the original tree
+                int indexOfLastLevel = dimensionNames.indexOf(levels.get(levels.size() - 1));
+                boolean isLeafNode = depth == indexOfLastLevel; // If this is the last level we aggregate, the new node should be a leaf
+                                                                // node
+                nodeInNewTree = new MDCSDimensionNode(dimensionValue, !isLeafNode, currentInOriginalTree.getStats());
+                parentInNewTree.children.put(dimensionValue, nodeInNewTree);
+            } else {
+                // Otherwise increment existing stats
+                CacheStatsCounterSnapshot newStats = CacheStatsCounterSnapshot.addSnapshots(
+                    nodeInNewTree.getStats(),
+                    currentInOriginalTree.getStats()
+                );
+                nodeInNewTree.setStats(newStats);
+            }
+            // Finally set the parent node to be this node for the next callers of this function
+            parentInNewTree = nodeInNewTree;
+        }
+
+        if (!currentInOriginalTree.children.isEmpty()) {
+            // Not a leaf node
+            for (Map.Entry<String, MDCSDimensionNode> childEntry : currentInOriginalTree.children.entrySet()) {
+                MDCSDimensionNode child = childEntry.getValue();
+                aggregateByLevelsHelper(parentInNewTree, child, levels, depth + 1);
+            }
+        }
+    }
```

Comment:
I am not exactly clear where the aggregateByLevels method is being used? Also, I am wondering if we create the MCDS directly on top from StatsHolder for the specific stats requested by user. If I understand correctly, currently we are cloning as immutable snapshot first and then pruning the level not needed

Reply:
aggregateByLevels() is used in toXContent(), which is in the upcoming PR. It could make sense to move it to that PR - let me know what you think. Unfortunately we do have to clone all snapshot values for all levels when we create the MCDS, because when we create it, we don't know which levels are going to be requested. The call to toXContent happens later, which means we have to prepare all possible levels to be able to return whatever levels are requested.